lib/core/src/Cardano/Wallet.hs

```diff
@@ -776,61 +778,58 @@ restoreWallet
     -> ExceptT ErrNoSuchWallet IO ()
 restoreWallet ctx wid = db & \DBLayer{..} -> do
     cps <- liftIO $ atomically $ listCheckpoints (PrimaryKey wid)
-    let forward bs h = run $ do
-            restoreBlocks @ctx @s @k ctx wid bs h
-    liftIO (follow nw tr cps forward (view #header)) >>= \case
-        FollowFailure ->
-            restoreWallet @ctx @s @k ctx wid
-        FollowRollback point -> do
-            rollbackBlocks @ctx @s @k ctx wid point
-            restoreWallet @ctx @s @k ctx wid
-        FollowDone ->
-            pure ()
-
+    let forward bs h innerTr = run $ do
+            restoreBlocks @ctx @s @k ctx innerTr wid bs h
+    liftIO (follow nw tr cps forward (view #header)) >>= \(innerTr, act) -> do
+        case act of
+            FollowFailure ->
+                restoreWallet @ctx @s @k ctx wid
+            FollowRollback point -> do
+                rollbackBlocks @ctx @s @k ctx innerTr wid point
+                restoreWallet @ctx @s @k ctx wid
+            FollowDone ->
+                pure ()
   where
     db = ctx ^. dbLayer @s @k
     nw = ctx ^. networkLayer
-    tr = contramap MsgFollow (ctx ^. logger @WalletLog)
+    tr = contramap MsgFollow (ctx ^. logger @WalletWorkerLog)
 
     run :: ExceptT ErrNoSuchWallet IO () -> IO (FollowAction ErrNoSuchWallet)
     run = fmap (either ExitWith (const Continue)) . runExceptT
 
 -- | Rewind the UTxO snapshots, transaction history and other information to a
 -- the earliest point in the past that is before or is the point of rollback.
 rollbackBlocks
-    :: forall ctx s k.
-        ( HasLogger WalletLog ctx
-        , HasDBLayer s k ctx
-        )
+    :: forall ctx s k. (HasDBLayer s k ctx)
     => ctx
+    -> Tracer IO WalletFollowLog
     -> WalletId
     -> SlotNo
     -> ExceptT ErrNoSuchWallet IO ()
-rollbackBlocks ctx wid point = db & \DBLayer{..} -> do
+rollbackBlocks ctx tr wid point = db & \DBLayer{..} -> do
```

Comment:
Just wondering out loud ... Perhaps we could even just remove the Tracer from `ctx`, and explicitly pass contramapped tracers around.

Reply:
🤔 I guess it would make it more difficult for functions in the Server module to call the functions here